pkg/core/tekton/tekton.go

```diff
@@ -0,0 +1,403 @@
+}
+
+// CreatePipeline receives a FuseML workflow and creates a Tekton pipeline from it
+func (t *WorkflowBackend) CreatePipeline(ctx context.Context, logger *log.Logger, workflow *workflow.Workflow) error {
+	pipeline := generatePipeline(*workflow, t.namespace)
+	logger.Printf("Creating tekton pipeline for workflow: %s...", workflow.Name)
+	_, err := t.tektonClients.PipelineClient.Create(ctx, pipeline, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RunPipeline creates a PipelineRun with its default values for received workflow and codeset
+func (t *WorkflowBackend) RunPipeline(ctx context.Context, workflowName string, codeset codeset.Codeset) error {
+	pipeline, err := t.tektonClients.PipelineClient.Get(ctx, workflowName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	pipelineRun, err := generatePipelineRun(pipeline, codeset)
+	if err != nil {
+		return err
+	}
+
+	_, err = t.tektonClients.PipelineRunClient.Create(ctx, pipelineRun, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// CreateListener creates tekton resources required to have a listener ready for triggering the pipeline
+func (t *WorkflowBackend) CreateListener(ctx context.Context, logger *log.Logger, workflowName string) (string, error) {
+
+	pipeline, err := t.tektonClients.PipelineClient.Get(ctx, workflowName, metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	triggerTemplate := generateTriggerTemplate(pipeline)
+	_, err = t.tektonClients.TriggerTemplateClient.Get(ctx, workflowName, metav1.GetOptions{})
+	if err != nil {
+		if !k8serr.IsNotFound(err) {
+			return "", err
+		}
+		logger.Printf("Creating tekton trigger template for workflow: %s...", workflowName)
+		_, err := t.tektonClients.TriggerTemplateClient.Create(ctx, triggerTemplate, metav1.CreateOptions{})
+		if err != nil {
+			return "", err
+		}
+	}
+
+	triggerBinding := generateTriggerBinding(triggerTemplate)
+	_, err = t.tektonClients.TriggerBindingClient.Get(ctx, workflowName, metav1.GetOptions{})
+	if err != nil {
+		if !k8serr.IsNotFound(err) {
+			return "", err
+		}
+		logger.Printf("Creating tekton trigger binding for workflow: %s...", workflowName)
+		_, err := t.tektonClients.TriggerBindingClient.Create(ctx, triggerBinding, metav1.CreateOptions{})
+		if err != nil {
+			return "", err
+		}
+	}
+
+	eventListener := generateEventListener(triggerTemplate, triggerBinding)
+	el, err := t.tektonClients.EventListenerClient.Get(ctx, workflowName, metav1.GetOptions{})
+	if err != nil {
+		if !k8serr.IsNotFound(err) {
+			return "", err
+		}
+		logger.Printf("Creating tekton event listener for workflow: %s...", workflowName)
+		el, err = t.tektonClients.EventListenerClient.Create(ctx, eventListener, metav1.CreateOptions{})
+		if err != nil {
+			return "", err
+		}
+	}
+
+	if err := waitFor(t.eventListenerReady(ctx, el.Name), 1*time.Second, 1*time.Minute); err != nil {
```

Comment:
Could you put the actual values into some variable?

Reply:
done